mlflow/pyfunc/agent_server/__init__.py

```diff
@@ -0,0 +1,485 @@
+    ResponsesAgentRequest,
+    ResponsesAgentResponse,
+    ResponsesAgentStreamEvent,
+)
+from mlflow.utils.annotations import experimental
+
+__all__ = [
+    "set_request_headers",
+    "get_request_headers",
+    "get_forwarded_access_token",
+    "get_header",
+    "get_obo_workspace_client",
+    "AgentServer",
+    "invoke",
+    "stream",
+    "parse_server_args",
+    "setup_mlflow",
+]
+
+
+_invoke_function: Callable[..., Any] | None = None
+_stream_function: Callable[..., Any] | None = None
+AgentType = Literal["agent/v1/responses", "agent/v1/chat", "agent/v2/chat"]
+
+
+@experimental(version="3.6.0")
+def invoke() -> Callable[[Callable[..., Any]], Callable[..., Any]]:
+    """Decorator to register a function as an invoke endpoint. Can only be used once."""
+
+    def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
+        global _invoke_function
+        if _invoke_function is not None:
+            raise ValueError("invoke decorator can only be used once")
+        _invoke_function = func
+        return func
+
+    return decorator
+
+
+def get_invoke_function():
+    return _invoke_function
+
+
+@experimental(version="3.6.0")
+def stream() -> Callable[[Callable[..., Any]], Callable[..., Any]]:
+    """Decorator to register a function as a stream endpoint. Can only be used once."""
+
+    def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
+        global _stream_function
+        if _stream_function is not None:
+            raise ValueError("stream decorator can only be used once")
+        _stream_function = func
+        return func
+
+    return decorator
+
+
+@experimental(version="3.6.0")
+class AgentValidator:
+    def __init__(self, agent_type: AgentType | None = None):
+        self.agent_type = agent_type
+        self.logger = logging.getLogger(__name__)
+
+    def validate_pydantic(self, pydantic_class: type[BaseModel], data: Any) -> None:
+        """Generic pydantic validator that throws an error if the data is invalid"""
+        if isinstance(data, pydantic_class):
+            return
+        try:
+            if isinstance(data, BaseModel):
+                pydantic_class(**data.model_dump())
+                return
+            pydantic_class(**data)
+        except Exception as e:
+            raise ValueError(
+                f"Invalid data for {pydantic_class.__name__} (agent_type: {self.agent_type}): {e}"
+            )
+
+    def validate_dataclass(self, dataclass_class: Any, data: Any) -> None:
+        """Generic dataclass validator that throws an error if the data is invalid"""
+        if isinstance(data, dataclass_class):
+            return
+        try:
+            dataclass_class(**data)
+        except Exception as e:
+            raise ValueError(
+                f"Invalid data for {dataclass_class.__name__} (agent_type: {self.agent_type}): {e}"
+            )
+
+    def validate_and_convert_request(self, data: dict[str, Any]) -> None:
```

Comment:
nit: return type annotation seems incorrect here - should it be something like:
`ResponsesAgentRequest | ChatCompletionRequest | ChatAgentRequest:`